examples/gno.land/r/gov/dao/types.gno

```diff
@@ -0,0 +1,227 @@
+
+func (pid ProposalID) String() string {
+	return ufmt.Sprintf("%v", int64(pid))
+}
+
+// VoteOption is the limited voting option for a DAO proposal
+// New govDAOs can create their own VoteOptions if needed in the
+// future.
+type VoteOption string
+
+const (
+	AbstainVote VoteOption = "ABSTAIN" // Side is not chosen
+	YesVote     VoteOption = "YES"     // Proposal should be accepted
+	NoVote      VoteOption = "NO"      // Proposal should be rejected
+)
+
+type VoteRequest struct {
+	Option     VoteOption
+	ProposalID ProposalID
+	Metadata   interface{}
+}
+
+func NewProposalRequest(title string, description string, executor Executor) ProposalRequest {
+	return ProposalRequest{
+		title:       title,
+		description: description,
+		executor:    executor,
+	}
+}
+
+func NewProposalRequestWithFilter(title string, description string, executor Executor, filter Filter) ProposalRequest {
+	return ProposalRequest{
+		title:       title,
+		description: description,
+		executor:    executor,
+		filter:      filter,
+	}
+}
+
+type Filter interface{}
+
+type ProposalRequest struct {
+	title       string
+	description string
+	executor    Executor
+	filter      Filter
+}
+
+func (p *ProposalRequest) Title() string {
+	return p.title
+}
+
+func (p *ProposalRequest) Description() string {
+	return p.description
+}
+
+func (p *ProposalRequest) Filter() Filter {
+	return p.filter
+}
+
+type Proposal struct {
+	author std.Address
+
+	title       string
+	description string
+
+	executor    Executor
+	allowedDAOs []string
+}
+
+func (p *Proposal) Author() std.Address {
+	return p.author
+}
+
+func (p *Proposal) Title() string {
+	return p.title
+}
+
+func (p *Proposal) Description() string {
+	return p.description
+}
+
+func (p *Proposal) ExecutorString() string {
+	if p.executor != nil {
+		return p.executor.String()
+	}
+
+	return ""
+}
+
+func (p *Proposal) AllowedDAOs() []string {
+	return p.allowedDAOs
+}
+
+type Proposals struct {
+	seq       seqid.ID
+	*avl.Tree // *avl.Tree[ProposalID]*Proposal
+}
+
+func NewProposals() *Proposals {
+	return &Proposals{Tree: avl.NewTree()}
+}
+
+func (ps *Proposals) SetProposal(p *Proposal) ProposalID {
+	pid := ProposalID(int64(ps.seq))
+	updated := ps.Set(pid.String(), p)
+	if updated {
+		panic("fatal error: Override proposals is not allowed")
+	}
+
+	ps.seq = ps.seq.Next()
+
+	return pid
+}
+
+func (ps *Proposals) GetProposal(pid ProposalID) *Proposal {
+	pv, ok := ps.Get(pid.String())
+	if !ok {
+		return nil
+	}
+
+	return pv.(*Proposal)
+}
+
+type Executor interface {
+	Execute() error
+	String() string
+}
+
+func NewSimpleExecutor(callback func() error, description string) *SimpleExecutor {
+	return &SimpleExecutor{
+		callback: callback,
+		desc:     description,
+	}
+}
+
+// SimpleExecutor implements the Executor interface using
+// a callback function and a description string.
+type SimpleExecutor struct {
+	callback func() error
+	desc     string
+}
+
+func (e *SimpleExecutor) Execute() error {
```

Comment:
This should be protected so that only a DAO can call Execute() after the voting process.

We cannot safely provide a pointer to an executor because anyone could execute it without going through the DAO voting system.

For composition, we want contracts to return safe executors, not proposals.

You can either wrap the callback in NewSimpleExecutor to ensure the callback is executed after a sanity check or modify the Execute function to include additional checks.

(Note that we may also want an "ExecuteMaximumOnce" check in the same wrapping.)

Reply:
Contracts are not returning safe proposals, they are returning safe ProposalRequests.

If needed, someone can add more security to an executor. It is composable enough.